File: mobile/src/api/groups.ts
import { apiClient } from './client';

export interface Group {
  id: string;
  name: string;
  description: string | null;
  avatar_url: string | null;
  owner_id: string;
  created_at: string;
}
export interface GroupMember {
  user_id: string;
  role: 'owner' | 'admin' | 'member';
}

export const listGroups = () => apiClient.get<Group[]>('/groups').then(r => r.data);
export const getGroup = (id: string) => apiClient.get<Group>(`/groups/${id}`).then(r => r.data);
export const createGroup = (name: string, description?: string) =>
  apiClient.post<Group>('/groups', { name, description: description ?? null }).then(r => r.data);
export const updateGroup = (id: string, patch: Partial<Pick<Group,'name'|'description'|'avatar_url'>>) =>
  apiClient.patch<Group>(`/groups/${id}`, patch).then(r => r.data);
export const listMembers = (id: string) =>
  apiClient.get<GroupMember[]>(`/groups/${id}/members`).then(r => r.data);
export const setMemberRole = (gid: string, uid: string, role: 'admin'|'member') =>
  apiClient.patch(`/groups/${gid}/members/${uid}/role`, { role });
export const kickMember = (gid: string, uid: string) =>
  apiClient.delete(`/groups/${gid}/members/${uid}`);
export const transferOwner = (gid: string, newOwnerId: string) =>
  apiClient.post(`/groups/${gid}/transfer`, { new_owner_id: newOwnerId });

File: backend/app/routers/groups.py
from fastapi import APIRouter, Depends, status
from pydantic import BaseModel

from app.dependencies.auth import CurrentUser, current_user
from app.dependencies.permissions import require_group_role
from app.models.group import (
    Group, GroupCreate, GroupMember, GroupUpdate, RoleUpdate,
)
from app.services.group_service import (
    create_group, get_group, list_members, list_my_groups,
    remove_member, set_role, transfer_owner, update_group,
)

router = APIRouter(prefix="/api/groups", tags=["groups"])


class TransferBody(BaseModel):
    new_owner_id: str


@router.post("", response_model=Group)
def create(body: GroupCreate, user: CurrentUser = Depends(current_user)) -> Group:
    return create_group(user.id, body)


@router.get("", response_model=list[Group])
def list_(user: CurrentUser = Depends(current_user)) -> list[Group]:
    return list_my_groups(user.id)


@router.get("/{gid}", response_model=Group)
def get_(gid: str, _: None = require_group_role("member")) -> Group:
    return get_group(gid)


@router.patch("/{gid}", response_model=Group)
def patch_(gid: str, body: GroupUpdate, _: None = require_group_role("admin")) -> Group:
    return update_group(gid, body)


@router.get("/{gid}/members", response_model=list[GroupMember])
def members(gid: str, _: None = require_group_role("member")) -> list[GroupMember]:
    return list_members(gid)


@router.patch("/{gid}/members/{uid}/role", status_code=status.HTTP_204_NO_CONTENT)
def patch_role(gid: str, uid: str, body: RoleUpdate, _: None = require_group_role("admin")) -> None:
    set_role(gid, uid, body.role)


@router.delete("/{gid}/members/{uid}", status_code=status.HTTP_204_NO_CONTENT)
def kick(gid: str, uid: str, _: None = require_group_role("admin")) -> None:
    remove_member(gid, uid)


@router.post("/{gid}/transfer", status_code=status.HTTP_204_NO_CONTENT)
def transfer(
    gid: str,
    body: TransferBody,
    user: CurrentUser = Depends(current_user),
    _: None = require_group_role("owner"),
) -> None:
    transfer_owner(gid, user.id, body.new_owner_id)
